File: Binary Search/binary_search.py
"""
https://leetcode.com/problems/binary-search/description/

"""
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while l <= r: # "<=" covers cases where the list contains a single item
            m = (l + r)//2

            if nums[m] < target:
                l = m + 1
            elif nums[m] > target:
                r = m - 1
            else:
                return m
            
        return -1

File: Binary Search/binary_search.js
/*
https://leetcode.com/problems/binary-search/description/
*/

class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number}
     * Time complexity O(nlogn) Space complexity O(n)
     */
    search(nums, target) { 
        nums.sort((a, b) => a - b)
        let l = 0
        let r = nums.length - 1

        while ( l <= r){
            let m = Math.trunc((r+l)/2)

            if (nums[m] > target){
                r = m - 1
            } else if (nums[m] < target){
                l = m + 1
            } else {
                return m
            }
        }
        return -1
    }
}
